### PR TITLE
Disable `build-macos` and `build-windows` on PR

### PR DESCRIPTION
**Pull request by jentfoo**
This commit removes the `build-macos` and `build-windows` from the PR flow, instead delegating to the bypass job. These jobs still run before being accepted in the merge queue.

This means that failures in these two jobs may not be known until the merge queue. Although late issue discovery is not ideal, it is considered the most critical in ensuring that `master` remains stable.

After discussing with @r0mant and @russjones we believe this change is worth trying because:
* Currently MacOS builds are 31% of our Teleport Actions spend (~$3,400 / week)
* Windows builds are also significant at 13% (~$1,400 / week)
* There has been relatively few failures of these jobs (without other jobs also failing)
* Unlike unit tests, build failures are never intermittent so a failure is a clear sign of a fix needed
* We still will have verification that `master` remains healthy

As an additional minor optimization all bypass jobs were updated to run on `ubuntu-latest`.

2 to 4 weeks after this change is merged we plan to review our spend again to see if these changes are helping as expected, as well if any additional opportunities for savings present themselves.

### CODE REVIEW

**Review by zmb3 — APPROVED** [review]
Seems fine to me.

File: .github/workflows/build-macos-bypass.yaml [hunk 1]
```diff
@@ -40,7 +29,7 @@ on:
 jobs:
   build:
     name: Build on Mac OS
```
Comment (reedloden) [inline .github/workflows/build-macos-bypass.yaml:31]: Can we update this?

File: .github/workflows/build-windows-bypass.yaml [hunk 2]
```diff
@@ -38,7 +28,7 @@ on:
 jobs:
   build:
     name: Build on Windows
```
Comment (reedloden) [inline .github/workflows/build-windows-bypass.yaml:30]: Can we update this?

File: .github/workflows/build-macos.yaml
Comment (reedloden) [inline .github/workflows/build-macos.yaml:1]: Could we change `macos-12-xl` to `macos-latest-xl` while we're at it? Unrelated to this change, but I was going to do that anyway.

**Review by reedloden — APPROVED** [review]
Thanks! A few nits.

File: .github/workflows/build-macos-bypass.yaml (hunk 1 above)
Comment (zmb3) [reply .github/workflows/build-macos-bypass.yaml:31]: No, the name has to be the same as the real workflow to pass required checks and get PRs added to the merge queue.

**Review by r0mant — APPROVED** [review]

File: .github/workflows/build-macos-bypass.yaml (hunk 1 above)
Comment (reedloden) [reply .github/workflows/build-macos-bypass.yaml:31]: I thought the workflow name just needs to be the same (see line 11). Does the job name have to be the same as well?

Comment (r0mant) [reply .github/workflows/build-macos-bypass.yaml:31]: Yes, all job names need to be the same.

File: .github/workflows/build-macos.yaml
Comment (jentfoo) [reply .github/workflows/build-macos.yaml:1]: It does not look like `macos-latest-xl` is a valid tag: https://github.com/gravitational/teleport/actions/runs/4631137125/jobs/8193619941

I removed this commit from this PR

Comment (reedloden) [reply .github/workflows/build-macos.yaml:1]: hmm, I'll open a support ticket, as https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources says it should work.

**Comment by jentfoo** [comment]
Backport PR's:
https://github.com/gravitational/teleport/pull/24231
https://github.com/gravitational/teleport/pull/24232
https://github.com/gravitational/teleport/pull/24233

**Pull request merged**